### DISCUSSION TITLE
npm ERR! explore It doesn't look like nodejs-extensions is installed

### DISCUSSION THREAD

**Discussion by giantpenguin** [Lombiq/Helpful-Extensions#147] [Q&A]
I am now facing this error "npm ERR! explore It doesn't look like nodejs-extensions is installed" when try executing the "build" task in Lombiq.NodeJs.Extensions.Samples

I have configured Lombiq.NodeJs.Extensions.Samples follow [As a Git submodule](https://github.com/Lombiq/NodeJs-Extensions#as-a-git-submodule) then trying execute the build task through the Task Runner Explorer.

What Target, variable should I check to overcome this error? 
Please help!

**Comment by Piedone** [comment]
Just to rule out things, can you build the [[Lombiq's Open-Source Orchard Core Extensions repo](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions)? Be sure to adhere to this: https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions#:~:text=Since%20the%20extensions%20are%20included%20as%20git%20submodules and this: https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions#prerequisites

**Comment by Aysenur-Erkin** [comment]
The "npm ERR! explore It doesn't look like nodejs-extensions is installed" error occurs because your project is missing or improperly installed a Node.js package called nodejs-extensions. To resolve this navigate to your project directory, open a command prompt and run the command "npm install nodejs-extensions" to install the missing package. This may help resolve the error.

**Reply by giantpenguin** [reply]
Hi @Aysenur-Erkin,

The "nodejs-extensions" indeed is defined in the Lombiq.NodeJs.Extensions' package.json file ({
  "name": "nodejs-extensions",  "version": "0.0.1",), it is not a public package for installing through npm install nodejs-extensions.

Lombiq.NodeJs.Extensions.targets, line ~112:
    
    <Copy SourceFiles="$(NodeJsExtensionsNpmPackageSourcePath)/config/consumer/package.project.json"
          DestinationFiles="$(MSBuildProjectDirectory)/package.json"
          Condition="'$(ShouldConfigureESLintPerProject)' == 'true' AND !Exists('$(MSBuildProjectDirectory)/package.json')" />

So because of something incorrect the "nodejs-extensions" reference has been misleading.

**Comment by giantpenguin** [comment]
Thank @Piedone & @Aysenur-Erkin for your helpful response!

> Yes, I totally agree that the error massage itself implies the nodejs-extensions is missing, I will try to install the package.
> But my concerns are:
> - there is no-where in many Readme files states that this nodejs-extensions package need to be installed as a prerequisite, even the NodeJs-Extensions' Readme file only states that we can execute the custom tasks through the VS Task Runner Explorer without mentioning about this package (https://github.com/Lombiq/NodeJs-Extensions/#how-to-trigger-pipelines-on-demand), neither in the existing package.json files defines nodejs-extensions in its "dependencies" node => this did make make me **have a myths that the NodeJs-Extensions handles the dependency packages already**.
> - this nodejs-extensions package must be installed in the "node-modules" folder of the consumer project (in this case the NodeJs-Extensions.Samples project), so we need to have nodejs-extensions package install for every consumer projects => whether I did something wrong? Or is it the current way?
> 
> With the above concerns I thought I have a wrong configuration.

The above "in quote" is in a wrong way of thinking. The NodeJs-Extensions does make itself as "nodejs-extensions" package
```
    <!-- Here we link to Node.js Extensions as an npm package, which makes it available to the consumer project under
         its package name "nodejs-extensions". This needs the consumer to have a package.json file, so if there is none,
         we create one. -->
    <Copy SourceFiles="$(NodeJsExtensionsNpmPackageSourcePath)/config/consumer/package.project.json"
          DestinationFiles="$(MSBuildProjectDirectory)/package.json"
          Condition="'$(ShouldConfigureESLintPerProject)' == 'true' AND !Exists('$(MSBuildProjectDirectory)/package.json')" />
```
but because of a wrong configuration (unknown) causes the issue.

Your inputs are really welcome. Thank you!

**Reply by Piedone** [reply]
Yes, you don't need to manually install anything, and NE is not a standard packaged downloadable from NPM but a local NPM package (this is because it needs to be consumed both by Node as an NPM package and by .NET projects as a NuGet package).

**Comment by giantpenguin** [comment]
Hi All,

Finally I have the Lombiq.NodeJs.Extensions.Samples build successfully. 
Since we already have has a notice of not using the default Lombiq folders structure (https://github.com/Lombiq/NodeJs-Extensions/blob/dev/Readme.md#as-a-git-submodule) related to the ./src/Utilities/ path, we must also check and update this path in **/config/consumer/.eslintrc.global.js**, **config/consumer/.stylelintrc.global.js** files.

**Reply by Piedone** [reply]
I'm glad that resolved this. I'm not sure I understand what exactly you did with the two config JS files, could you please elaborate?

**Reply by giantpenguin** [reply]
So I try to learn the Lombiq.NodeJs.Extensions and some other Lombiq projects by project references but place all the Lombiq projects are in a same folder (in the same level - no Libraries, Utilities folders) so that I must update all the paths start with Libraries/, Utilities/ as mentioned in https://github.com/Lombiq/NodeJs-Extensions/blob/dev/Readme.md#as-a-git-submodule.
The 2files /config/consumer/.eslintrc.global.js, config/consumer/.stylelintrc.global.js also have path start with ./src/Utilities/ that need to be updated align with the folder structure changed.

Thank you, @Piedone!

**Reply by Piedone** [reply]
Ah I see. Do you think our documentation could be clear on this aspect? If yes, would you venture into suggesting changes, even as awesome as a pull request?

**Reply by giantpenguin** [reply]
Hi @Piedone,

I think the document should include a notice about the 2 files /config/consumer/.eslintrc.global.js, config/consumer/.stylelintrc.global.js also need to be updated in case the relevant paths (the folder structure) has been changed.

Indeed the document and comments in the project are so helpful, but the project itself is so complicated then the document has been kept clean as the current is best.

Thanks & regards,

**Reply by Piedone** [reply]
How does this sound? https://github.com/Lombiq/NodeJs-Extensions/blob/dev/Readme.md#as-a-git-submodule:~:text=in%20that%20case%20you%27ll%20need%20to%20adjust%20the%20paths

**Comment by giantpenguin** [comment]
Yeah, that is a really neat update ;)

**Reply by Piedone** [reply]
Great then!
